libsolidity/experimental/ast/TypeSystem.h

```diff
@@ -105,7 +105,7 @@ class TypeSystem
 
 	struct TypeClassInfo
 	{
-		Type typeVariable;
+		TypeVariable typeVariable;
```

Comment:
Don't do this. ``TypeVariable`` is not meant to be external interface for various reasons like not being default-constructible.

Reply:
``TypeVariables`` are intentionally not constructible. They're an implementation detail of types. Mixing them into the external interface or making them visible on the user-side is only bound to cause problems.

Reply:
I find it really odd that as a user you never want to speak about type variables and distinguish them from types. Having to treat it as an opaque object and not look inside sounds reasonable, but I think that being able to know that you have specifically a variable, and not an arbitrary type is a useful distinction in the interface. IMO hiding it will create more nuisance than it solves real problems.

Reply:
The main interface for types is unification. If you start to manually distinguish between type variables and other types you're in almost all cases doing something wrong. You may want to deconstruct specific type constants, but there is helper interfaces for that.

Reply:
I don't really buy this: the current code cares about type variables in a lot of places. For example when validating type variables in type class definitions. Or when fixing type variables. I really don't like the weird patterns I have to resort to to get at the type variable in my PRs when it could be so straightforward.

But anyway, perhaps this is not the best place to continue this discussion. I'll resubmit this later in a PR, when my fixed variable PRs are merged and maybe you'll change your mind when you see the cruft this would remove.